nova/virt/vmwareapi/driver.py

```diff
@@ -1022,12 +1023,28 @@ def rollback_live_migration_at_destination(self, context, instance,
         if not migrate_data.is_same_vcenter:
             self._volumeops.delete_shadow_vms(block_device_info, instance)
 
+    @contextlib.contextmanager
+    def _error_out_instance_on_exception(self, instance, message):
+        try:
+            yield
+        except Exception as error:
+            LOG.exception("Failed to %s, setting to ERROR state",
+                          message,
+                          instance=instance, error=error)
+            instance.vm_state = vm_states.ERROR
+            instance.save()
```

Comment:
We only error out the instance, while the migration would report success. Would it make sense to also error out the migration?
```
                # Put instance and migration into error state,
                # as its almost certainly too late to rollback
                self._set_migration_status(migration, 'error')
``` (https://github.com/sapcc/nova/blob/6b46ce32cab35211d2ac6b6e757ecb5f95dd0d90/nova/compute/manager.py#L6776-L6778)

Reply:
Certainly, it is an option to set the migration as failed to failed to inform the user (or better said admin) about the failure.
But from my point-of-view, the migration did succeed, we "only" fail in the cleanup. So the host should be set to the new hypervisor, etc..

We need to error out the instance, so the user cannot modify the instance and more importantly the attached volumes, so the user cannot modify it without.
Failing the migration requires larger changes, and is purely informational, and possibly breaking the "interface". I would expect the VM after a failed migration on the original hypervisor.

Reply:
hm ... the comment from Nova looks like they would also set a migration to failed, even if they can't roll back. So I don't think your expectation matches upstream's expectation. I think every time the migration failed at some point, the migration is errored out - there's no assumption made over where the VM is running then, because something unexpected happened ;)

That being said, erroring out the migration does not help in guarding the instance and I don't feel too strongly about this. I would just have preferred to see if this was a clean migration or if errors happened during the migration. Might get helpful to find these problems automatically.

Btw. a user can reset the state to ACTIVE at least via dashboard - and they do tend to use that feature if problems occur. Is there another state or maybe task_state we could set, so that it's clear that this instance is off limits to users for a time? Something like "in maintenance" or so, that only we can clear?

Reply:
Resetting the state of a VM from an ERROR state is an admin action for a reason. It is the state the VM lands in, if an operator needs to have a look at it.
If people do that deliberately over the dashboard, I would suggest to put a disclaimer there, that it might seriously break things with data-loss. This is not limited to this change.

I do not find any "maintenance" state for the VM: https://github.com/sapcc/nova/blob/stable/rocky-m3/nova/objects/fields.py#L879-L890

Reply:
Yeah, there's nothing in there. Only thing we could try is to keep the `task_state` on `migrating` or something. Not sure if that even helps.

I just fear that there will no operator available in a timely manner and people will just change stuff, because we want to hand out the live-migration endpoint to automation for VM scheduling.

Reply:
Well, hopefully this never is triggered, or at the very least doesn't end up in a state, which causes data loss.
But we really need to educate the users about the risks of resetting the error state.
My understanding is that we handed it out to admin users of the customers because they should be able to make an informed decision themselves, and are not blocked by an operator. I still think that is valid position.

The second part is we really should then look into our alerting and improve our responsiveness probably by cutting down the noise before making this fully automated.